common/src/chain/transaction/signature/inputsig/authorize_pubkeyhash_spend.rs

```diff
@@ -77,4 +77,223 @@ pub fn sign_address_spending(
+            signature::{
+                inputsig::{InputWitness, StandardInputSignature},
+                sighashtype::SigHashType,
+                signature_hash,
+            },
+            Destination, Transaction, TransactionCreationError, TxInput, TxOutput,
+        },
+        primitives::{Amount, Id},
+    };
+
+    use super::*;
+
+    fn generate_unsigned_tx(
+        outpoint_dest: Destination,
+    ) -> Result<Transaction, TransactionCreationError> {
+        let tx = Transaction::new(
+            0,
+            vec![TxInput::new(
+                Id::<Transaction>::new(&H256::zero()).into(),
+                0,
+                InputWitness::NoSignature(None),
+            )],
+            vec![TxOutput::new(Amount::from_atoms(100), outpoint_dest)],
+            0,
+        )?;
+        Ok(tx)
+    }
+
+    fn prepare_data_for_verify(
+        sighash_type: SigHashType,
+    ) -> (PublicKeyHash, AuthorizedPublicKeyHashSpend, H256) {
+        const INPUT_NUM: usize = 0;
+        let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let pubkey_hash = PublicKeyHash::from(&public_key);
+        let outpoint_dest = Destination::Address(pubkey_hash);
+        let tx = generate_unsigned_tx(outpoint_dest.clone()).unwrap();
+        let witness = StandardInputSignature::produce_signature_for_input(
+            &private_key,
+            sighash_type,
+            outpoint_dest,
+            &tx,
+            INPUT_NUM,
+        )
+        .unwrap();
+        let spender_signature =
+            AuthorizedPublicKeyHashSpend::from_data(witness.get_raw_signature()).unwrap();
+        let sighash = signature_hash(witness.sighash_type(), &tx, INPUT_NUM).unwrap();
+        (pubkey_hash, spender_signature, sighash)
+    }
+
+    fn prepare_data_for_wrong_destination(sighash_type: SigHashType) -> StandardInputSignature {
+        const INPUT_NUM: usize = 0;
+        let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let outpoint_dest = Destination::PublicKey(public_key);
+        let tx = generate_unsigned_tx(outpoint_dest.clone()).unwrap();
+        StandardInputSignature::produce_signature_for_input(
+            &private_key,
+            sighash_type,
+            outpoint_dest,
+            &tx,
+            INPUT_NUM,
+        )
+        .unwrap()
+    }
+
+    #[test]
+    fn test_wrong_destination() {
+        use std::mem::discriminant;
+        // Destination = Address, but we try to use here AuthorizedPublicKeyHashSpend
+        let sighash_type = SigHashType::try_from(SigHashType::ALL).unwrap();
+        let witness = prepare_data_for_wrong_destination(sighash_type);
+        // Compare Err(TransactionSigError) without inner message
+        assert_eq!(
+            discriminant(&AuthorizedPublicKeyHashSpend::from_data(
+                witness.get_raw_signature()
+            )),
+            discriminant(&Err(TransactionSigError::AddressAuthDecodingFailed(
```

Comment:
Please don't use discriminant. In case you're forced to compare enum arms, use `matches!` macro instead. Something like this:

```
        matches!(
            AuthorizedPublicKeyHashSpend::from_data(witness.get_raw_signature()).unwrap_err(),
            TransactionSigError::AddressAuthDecodingFailed(_)
        );
```

Reply:
I didn't know that matches! can do it. Thanks! Fixed!